staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/start.go

```diff
@@ -92,8 +92,11 @@ func (o *CustomResourceDefinitionsServerOptions) Complete() error {
 }
 
 func (o CustomResourceDefinitionsServerOptions) Config() (*apiserver.Config, error) {
-	// TODO have a "real" external address
-	if err := o.RecommendedOptions.SecureServing.MaybeDefaultWithSelfSignedCerts("localhost", nil, []net.IP{net.ParseIP("127.0.0.1")}); err != nil {
+	advertiseAddr, err := o.RecommendedOptions.SecureServing.DefaultExternalAddress()
+	if err != nil {
+		return nil, err
+	}
+	if err := o.RecommendedOptions.SecureServing.MaybeDefaultWithSelfSignedCerts(advertiseAddr.String(), nil, []net.IP{net.ParseIP("127.0.0.1")}); err != nil {
```

Comment:
is it the same we do in kube-apiserver?

Reply:
Here is kube-apiserver usage,  similar way to get public address.
```
	if err := s.SecureServing.MaybeDefaultWithSelfSignedCerts(s.GenericServerRunOptions.AdvertiseAddress.String(), []string{"kubernetes.default.svc", "kubernetes.default", "kubernetes"}, []net.IP{apiServerServiceIP}); err != nil {
		return fmt.Errorf("error creating self-signed certificates: %v", err)
	}
```

Reply:
If you pass an external address on the command line, does this honor it or still try to do auto detection?

Reply:
Honor it. 
```
func (s *SecureServingOptions) DefaultExternalAddress() (net.IP, error) {
	return utilnet.ChooseBindAddress(s.BindAddress)
}

func ChooseBindAddress(bindAddress net.IP) (net.IP, error) {
	if bindAddress == nil || bindAddress.IsUnspecified() || bindAddress.IsLoopback() {
		hostIP, err := ChooseHostInterface()
		if err != nil {
			return nil, err
		}
		bindAddress = hostIP
	}
	return bindAddress, nil
}
```

Reply:
Should we re-add localhost? Or was it removed intentionally?

Reply:
I think it is removed intentionally in kube-apiserver, so I just copy it.

Reply:
in the default case of listening to 0.0.0.0, do we get localhost automatically? Or what is the exernal address in that situation?

Reply:
This way it get address by  `ChooseHostInterface` from route file firstly if it  exists,  otherwise from nic.

Reply:
We loose "localhost" as a valid ServerName here. We should at least add it to the alternative DNS names, shouldn't we? Otherwise, people with `curl localhost` will get a TLS error after this.

Reply:
curl localhost == curl `127.0.0.1`, am I right?

Reply:
No, `curl localhost`, i.e. the http ServerName is `localhost`.

It's not really critical as we talk about apiservers that nobody runs outside of kube-apiserver. So maybe let's just ignore my comment and align it with kube-apiserver (which already ignores `localhost`).

Reply:
No, I mean the left one, i.e. ServerName `localhost`.

Reply:
don't remember the IP logic of HTTP clients though: will they be happy with domains resolving to an IP that is in the cert?

Reply:
IIRC, server use subjectAltName to check ServerName in headers.

Reply:
> will they be happy with domains resolving to an IP that is in the cert?

No

Reply:
Then pass `localhost` to the second param `alternateDNS` is more secure way.